src/main/kotlin/org/rust/lang/core/psi/ext/RsQualifiedNamedElement.kt

```diff
@@ -14,3 +19,162 @@ val RsQualifiedNamedElement.qualifiedName: String? get() {
     val cargoTarget = containingCargoTarget?.normName ?: return null
     return "$cargoTarget$inCratePath"
 }
+
+class RsQualifiedName private constructor(
+    private val crateName: String,
+    private val modSegments: List<String>,
+    private val parentItem: Item,
+    private val childItem: Item?
+) {
+
+    fun toUrlPath(): String {
+        val segments = mutableListOf(crateName)
+        segments += modSegments
+        val (pageName, anchor) =
+        if (parentItem.type == MOD) {
+            "index.html" to ""
+        } else {
+            "$parentItem.html" to (if (childItem != null) "#$childItem" else "")
+        }
+        segments += pageName
+        return segments.joinToString(separator = "/", postfix = anchor)
+    }
+
+    companion object {
+
+        @JvmStatic
+        fun from(element: RsQualifiedNamedElement): RsQualifiedName? {
+
+            val parent = parentItem(element)
+
+            val (parentItem, childItem) = if (parent != null) {
+                parent to (element.toChildItem() ?: return null)
+            } else {
+                val parentItem = element.toParentItem() ?: return null
+                parentItem to null
+            }
+
+            val crateName = element.containingCargoTarget?.normName ?: return null
+            val modSegments = if (parentItem.type == ParentItemType.PRIMITIVE) {
+                listOf()
+            } else {
+                element.containingMod.superMods
+                    .asReversed()
+                    .drop(1)
+                    .map { it.modName ?: return null }
+            }
+
+            return RsQualifiedName(crateName, modSegments, parentItem, childItem)
+        }
+
+        private fun parentItem(element: RsQualifiedNamedElement): Item? {
+            val parentItem: RsQualifiedNamedElement? = when (element) {
+                is RsAbstractable -> {
+                    val owner = element.owner
+                    when (owner) {
+                        is RsAbstractableOwner.Trait -> owner.trait
+                        is RsAbstractableOwner.Impl -> {
+                            return owner.impl.typeReference?.toParentItem()
+                        }
+                        else -> return null
+                    }
+                }
+                is RsEnumVariant -> element.parentEnum
+                is RsFieldDecl -> element.parentStruct
```

Comment:
Should we take into account that enums can contain fields too?

Reply:
It's an interesting question
On the one hand, named field in enum variant can have qualified name
On the other hand, rustdoc doesn't add info about it in a documentation
So, let's do same things as rustdoc and don't generate qualified name for fields in enum variant now and will implement it later if it is necessary